ansible_collections/arista/avd/plugins/filter/add_md_toc.py

```diff
@@ -110,27 +104,74 @@ def add_md_toc(md_input, skip_lines=0, toc_levels=3, toc_marker="<!-- toc -->"):
         MD with added TOC
     """
 
-    if isinstance(md_input, Undefined) or md_input is None or HAS_MD_TOC is False:
-        # Return None
-        return
-
-    # Generate TOC from variable
-    with StringIO(md_input) as md:
-        stdin = sys.stdin
-        sys.stdin = md
-        try:
-            # Try using new md_toc api when md-toc>=9.0.0.
-            toc = md_toc.api.build_toc("-", keep_header_levels=toc_levels, skip_lines=skip_lines).rstrip()
-        except AttributeError:
-            # If that fails, use the previous version md-toc>=8.1.0,<9.0.0
-            toc = md_toc.build_toc("-", keep_header_levels=toc_levels, skip_lines=skip_lines).rstrip()
-        sys.stdin = stdin
-
-    # Insert TOC between markers
-    toc_marker = re.escape(toc_marker)
-    toc_pattern = re.compile(rf"{toc_marker}[\S\s]*?{toc_marker}")
-
-    return toc_pattern.sub(toc, md_input, count=1)
+    if not isinstance(skip_lines, int):
+        raise AnsibleFilterError(f"add_md_toc 'skip_lines' argument must be an integer. Got '{skip_lines}'({type(skip_lines)}).")
+
+    if not isinstance(toc_levels, int) or toc_levels < 1:
+        raise AnsibleFilterError(f"add_md_toc 'toc_levels' argument must be >0. Got '{toc_levels}'({type(skip_lines)}).")
+
+    if not isinstance(toc_marker, str) or not toc_marker:
+        raise AnsibleFilterError(f"add_md_toc 'toc_marker' argument must be a non-empty string. Got '{toc_marker}'({type(skip_lines)}).")
+
+    if not isinstance(md_input, str):
+        raise AnsibleFilterError(f"add_md_toc expects a string. Got {type(md_input)}.")
+
+    md_lines = md_input.split("\n")
+    toc_marker_positions = []
+    toc_lines = []
+    all_anchor_ids = []
+
+    # toc_level_offset ensures we start the TOC at the lowest level within the unskipped lines.
+    toc_level_offset = 99
+
+    for line_num, line in enumerate(md_lines):
+        if line == toc_marker:
+            toc_marker_positions.append(line_num)
+            continue
+        if re.match(HEADING_PATTERN, line):
+            level, text, anchor_id = _get_line_info(line, all_anchor_ids)
+            if line_num < skip_lines or level > toc_levels:
```

Comment:
line_num check can be made before running the match on the regex

Reply:
No because we need the `all_anchor_ids` to be maintained even if not generating a TOC line. That is what caused the bug in md-toc which we are resolving with this logic.